src/Util/Memoize.php

```diff
@@ -0,0 +1,36 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Roave\BetterReflection\Util;
+
+/** @template T */
```

Comment:
I'm thinking of this, meanwhile:

```php
/** 
 * @template T
 * @internal do not touch: you have been warned.
 */
final class Memoize
{
    private readonly mixed $cached;
    /** @param pure-Closure(): T $cb */
    public function __construct(private Closure|null $cb) {}
    public function get(): mixed {
        if ($this->cb) {
            $this->cached = ($this->cb)();
            $this->cb = null;
       }
        
        return $this->cached;
    }
}
```

This could be used general-purpose on most adapters, if needed.

Reply:
copied it over. the SA tools don't like it ;)

Reply:
Well, that's good :D 

We are indeed bending the rules here: hence the `@internal` stuff as well.

I'll likely need to take over on this: a good measure of error suppression will be needed, but the `impure-closure` bit is most annoying (it shouldn't consider it impure :thinking: )